README


This archive contains the distribution MooseX-Types-URI,
version 0.07:

  URI related types and coercions for Moose

This software is copyright (c) 2008 by יובל קוג'מן (Yuval Kogman).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
Revision history for MooseX-Types-URI

0.07      2014-06-30 19:37:22Z
  - add missing prereq declaration

0.06      2014-06-28 17:00:56Z
  - better fix for fully-qualified type names, made to MooseX::Types directly
  - load URI::file, URI::data, URI::FromHash dynamically only when used
    (Olivier Mengué).

0.05      2013-12-17 23:20:10Z
  - fully-qualified type names like MooseX::Types::URI::Uri work again
    (removed in 0.04)

0.04      2013-11-15 03:50:07Z
  - convert to Dist::Zilla, fixing metadata
  - fix coercion from HashRef (thanks, MORIYA Masaki!) (RT#69476)

0.03  2011-11-17 15:40:20Z
  - Adapt to the new type constraint inlining API in Moose 2.0100
    (Shawn M Moore).

0.02  2009-05-27 17:49:49Z
  - Use MooseX::Types type declarators in the type coercions (Daniel Pittman).
  - Coerce from additional MooseX::Types::Path::Class types (Daniel Pittman).

0.01  2009-01-11 04:53:35Z
  - Initial release.


Build.PL
# This Build.PL for MooseX-Types-URI was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.005
if (eval 'use Module::Build::Tiny 0.036; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for MooseX-Types-URI was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.006.
    use 5.006;
    # use Module::Build::Tiny 0.036;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => "6.30",
        "Module::Build::Tiny" => "0.036"
      },
      "dist_abstract" => "URI related types and coercions for Moose",
      "dist_author" => [
        "\x{5d9}\x{5d5}\x{5d1}\x{5dc} \x{5e7}\x{5d5}\x{5d2}'\x{5de}\x{5df} (Yuval Kogman) <nothingmuch\@woobling.org>"
      ],
      "dist_name" => "MooseX-Types-URI",
      "dist_version" => "0.07",
      "license" => "perl",
      "module_name" => "MooseX::Types::URI",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "Moose::Util::TypeConstraints" => 0,
        "MooseX::Types" => "0.40",
        "MooseX::Types::Moose" => 0,
        "MooseX::Types::Path::Class" => 0,
        "Scalar::Util" => 0,
        "URI" => 0,
        "URI::FromHash" => 0,
        "URI::QueryParam" => 0,
        "URI::WithBase" => 0,
        "URI::data" => 0,
        "URI::file" => 0,
        "if" => 0,
        "namespace::autoclean" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "File::Spec::Functions" => 0,
        "List::Util" => 0,
        "Moose" => 0,
        "Test::More" => 0,
        "ok" => 0,
        "version" => 0
      }
    );


    my %fallback_build_requires = (
      "ExtUtils::MakeMaker" => 0,
      "File::Spec::Functions" => 0,
      "List::Util" => 0,
      "Moose" => 0,
      "Test::More" => 0,
      "ok" => 0,
      "version" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}


lib/MooseX/Types/URI.pm
package MooseX::Types::URI;
BEGIN {
  $MooseX::Types::URI::AUTHORITY = 'cpan:NUFFIN';
}
# git description: v0.05-16-g576eea8
$MooseX::Types::URI::VERSION = '0.06';
# ABSTRACT: URI related types and coercions for Moose
# KEYWORDS: moose types constraints coercions uri path web

use strict;
use warnings;

use Scalar::Util qw(blessed);

use URI;
use URI::QueryParam;
use URI::WithBase;

use Moose::Util::TypeConstraints;

use MooseX::Types::Moose qw{Str ScalarRef HashRef};
use MooseX::Types::Path::Class qw{File Dir};

use MooseX::Types 0.40 -declare => [qw(Uri _UriWithBase _Uri FileUri DataUri)];
use if MooseX::Types->VERSION >= 0.42, 'namespace::autoclean';

my $uri = Moose::Meta::TypeConstraint->new(
    name   => Uri,
    parent => Moose::Meta::TypeConstraint::Union->new(
        name => join("|", _Uri, _UriWithBase),
        type_constraints => [
            class_type( _Uri,         { class => "URI" } ),
            class_type( _UriWithBase, { class => "URI::WithBase" } ),
        ],
    ),
    ($Moose::VERSION >= 2.0100
        ? (inline_as => sub { 'local $@; blessed('.$_[1].') && ( '.$_[1].'->isa("URI") || '.$_[1].'->isa("URI::WithBase") )' })
        : (optimized => sub { local $@; blessed($_[0]) && ( $_[0]->isa("URI") || $_[0]->isa("URI::WithBase") ) })
    ),
);

register_type_constraint($uri);

coerce( Uri,
    from Str                 , via { URI->new($_) },
    from "Path::Class::File" , via { require URI::file; URI::file::->new($_) },
    from "Path::Class::Dir"  , via { require URI::file; URI::file::->new($_) },
    from File                , via { require URI::file; URI::file::->new($_) },
    from Dir                 , via { require URI::file; URI::file::->new($_) },
    from ScalarRef           , via { my $u = URI->new("data:"); $u->data($$_); $u },
    from HashRef             , via { require URI::FromHash; URI::FromHash::uri_object(%$_) },
);

class_type FileUri, { class => "URI::file", parent => $uri };

coerce( FileUri,
    from Str                 , via { require URI::file; URI::file::->new($_) },
    from File                , via { require URI::file; URI::file::->new($_) },
    from Dir                 , via { require URI::file; URI::file::->new($_) },
    from "Path::Class::File" , via { require URI::file; URI::file::->new($_) },
    from "Path::Class::Dir"  , via { require URI::file; URI::file::->new($_) },
);

class_type DataUri, { class => "URI::data" };

coerce( DataUri,
    from Str       , via { my $u = URI->new("data:"); $u->data($_);  $u },
    from ScalarRef , via { my $u = URI->new("data:"); $u->data($$_); $u },
);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Types::URI - URI related types and coercions for Moose

=head1 VERSION

version 0.06

=head1 SYNOPSIS

	use MooseX::Types::URI qw(Uri FileUri DataUri);

=head1 DESCRIPTION

This package provides Moose types for fun with L<URI>s.

=head1 TYPES

=head2 C<Uri>

Either L<URI> or L<URI::WithBase>

Coerces from C<Str> via L<URI/new>.

Coerces from L<Path::Class::File> and L<Path::Class::Dir> via L<URI::file/new>.

Coerces from C<ScalarRef> via L<URI::data/new>.

Coerces from C<HashRef> using L<URI::FromHash>.

=head2 C<DataUri>

A URI whose scheme is C<data>.

Coerces from C<Str> and C<ScalarRef> via L<URI::data/new>.

=head2 C<FileUri>

A L<URI::file> class type.

Has coercions from C<Str>, L<Path::Class::File> and L<Path::Class::Dir> via L<URI::file/new>

=for stopwords DWIMier ducktyping

It has slightly DWIMier types than the L<URI> classes have due to
implementation details, so the types should be more forgiving when ducktyping
will work anyway (e.g. L<URI::WithBase> does not inherit L<URI>).

=head1 TYPES

The types are with C<ucfirst> naming convention so that they don't mask the
L<URI> class.

=for stopwords TODO

=head1 TODO

Think about L<Path::Resource> integration of some sort

=head1 AUTHOR

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=head1 CONTRIBUTORS

=over 4

=item *

Daniel Pittman <daniel@rimspace.net>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

Karen Etheridge <ether@cpan.org>

=item *

MORIYA Masaki (gardejo) <moriya@ermitejo.com>

=item *

Olivier Mengué <dolmen@cpan.org>

=item *

Shawn M Moore <sartak@gmail.com>

=item *

Yuval Kogman <nothingmuch@woobling.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by יובל קוג'מן (Yuval Kogman).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This Makefile.PL for MooseX-Types-URI was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.010
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.20.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '6.30',
    'Module::Build::Tiny' => '0.036',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "URI related types and coercions for Moose",
  "AUTHOR" => "\x{5d9}\x{5d5}\x{5d1}\x{5dc} \x{5e7}\x{5d5}\x{5d2}'\x{5de}\x{5df} (Yuval Kogman) <nothingmuch\@woobling.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build::Tiny" => "0.036"
  },
  "DISTNAME" => "MooseX-Types-URI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Types::URI",
  "PREREQ_PM" => {
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Types" => "0.40",
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Class" => 0,
    "Scalar::Util" => 0,
    "URI" => 0,
    "URI::FromHash" => 0,
    "URI::QueryParam" => 0,
    "URI::WithBase" => 0,
    "URI::data" => 0,
    "URI::file" => 0,
    "if" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Moose" => 0,
    "Test::More" => 0,
    "ok" => 0,
    "version" => 0
  },
  "VERSION" => "0.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "List::Util" => 0,
  "Moose" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "MooseX::Types" => "0.40",
  "MooseX::Types::Moose" => 0,
  "MooseX::Types::Path::Class" => 0,
  "Scalar::Util" => 0,
  "Test::More" => 0,
  "URI" => 0,
  "URI::FromHash" => 0,
  "URI::QueryParam" => 0,
  "URI::WithBase" => 0,
  "URI::data" => 0,
  "URI::file" => 0,
  "if" => 0,
  "namespace::autoclean" => 0,
  "ok" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



